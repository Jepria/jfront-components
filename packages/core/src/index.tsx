export * from '@jfront-ui/app'
export * from '@jfront-ui/datepicker'
export * from '@jfront-ui/form'
//export * from '@jfront-ui/grid'
export * from '@jfront-ui/hooks'
//export * from '@jfront-ui/input'
export * from '@jfront-ui/layout'
//export * from '@jfront-ui/pagingbar'
//export * from '@jfront-ui/tabpanel'
//export * from '@jfront-ui/toolbar'
