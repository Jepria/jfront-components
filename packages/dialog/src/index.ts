export * from './SimpleDialog'
export * from './AlertDialog'
export * from './FormDialog'
export * from './ErrorDialog'
export * from './WarningDialog'