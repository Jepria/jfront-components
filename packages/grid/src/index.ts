export * from './common'
export * from './jepgrid'
