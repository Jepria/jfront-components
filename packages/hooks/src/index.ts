export * from "./useSelect"
export * from "./useSelect/useDropdown"
export * from "./useSelect/useDual"
export * from "./useSelect/useFilter"
export * from "./useSelect/useMultiple"
export * from "./useOnClickOutside"
export * from "./useRouteMapping"
