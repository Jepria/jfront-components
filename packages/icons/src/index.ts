export * from "./Add"
export * from "./Delete"
export * from "./Edit"
export * from "./Exclamation"
export * from "./First"
export * from "./Last"
export * from "./Loading"
export * from "./Next"
export * from "./OpenIcon"
export * from "./Prev"
export * from "./Refresh"
export * from "./Save"
export * from "./Search"
export * from "./SettingsSmall"
export * from "./Split"
export * from "./View"
