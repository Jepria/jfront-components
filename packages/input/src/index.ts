export interface InputProps {
  error?: string
  name?: string
  isLoading?: boolean
}

export * from "./SelectInput"
export * from "./TextInput"
export * from "./NumberInput"
export * from "./MaskedInput"
export * from "./MoneyInput"
export * from "./DecimalInput"
export * from "./MoneyInput"
