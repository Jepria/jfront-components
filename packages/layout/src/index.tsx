import styled from 'styled-components';

export const FormContainer = styled.div`
  padding: 5px;
`;

export const VerticalLayout = styled.div`
  width: 100%;
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-orient: vertical;
  -webkit-box-direction: normal;
      -ms-flex-direction: column;
          flex-direction: column;
`;

export const HorizontalLayout = styled.div`
  width: 100%;
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-orient: horizontal;
  -webkit-box-direction: normal;
      -ms-flex-direction: row;
          flex-direction: row;
`;

export const Page = styled.div`
  box-sizing: border-box;
  width: 100%;
  height: 100%;
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-orient: vertical;
  -webkit-box-direction: normal;
      -ms-flex-direction: column;
          flex-direction: column;
`;

export const Header = styled.header`
  box-sizing: border-box;
  height: auto;
  width: 100%;
`;

export const Content = styled.section`
  box-sizing: border-box;
  width: 100%;
  overflow: auto;
  -webkit-box-flex: 1;
      -ms-flex: 1;
          flex: 1;
  -ms-flex-positive: 1;
      flex-grow: 1;
`;

export const Footer = styled.footer`
  box-sizing: border-box;
  height: auto;
  width: 100%;
`;
