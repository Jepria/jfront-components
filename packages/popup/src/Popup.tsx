import React, {
  useState,
  useEffect,
  useRef,
  RefObject,
  useMemo,
  useLayoutEffect,
} from "react"
import { forwardRef } from "@jfront/ui-utils"
import {
  ElementPosition,
  useOnClickOutside,
  useResizePosition,
  useDragPosition,
  useScrollPosition,
} from "@jfront/ui-hooks"
import styled from "styled-components"
import ReactDOM from "react-dom"

const StyledDiv = styled.div`
  box-sizing: border-box;
  position: absolute;
  z-index: 5100;
  margin: 0;
  padding: 0;
  background: #fff;
  overflow: auto;
  box-shadow: 0px 5px 5px -3px rgba(0, 0, 0, 0.2),
    0px 8px 10px 1px rgba(0, 0, 0, 0.14), 0px 3px 14px 2px rgba(0, 0, 0, 0.12);
  outline: 0;
  min-width: 16px;
  min-height: 16px;
  overflow-x: hidden;
  overflow-y: auto;
`

export interface RelativePosition {
  horizontal?: "left" | "center" | "right"
  vertical?: "top" | "center" | "bottom"
}

export interface Position {
  top?: string
  left?: string
  right?: string
  bottom?: string
}

export interface PopupProps {
  className?: string
  style?: React.CSSProperties
  targetElementRef?: React.RefObject<HTMLElement>
  targetRelativePosition?: RelativePosition
  visible?: boolean
  children?: React.ReactNode
  disablePortal?: boolean
  flippingEnabled?: boolean
  onFlip?: (position: RelativePosition) => void
  onOpen?: () => void
  onClose?: () => void
}

const LEFT = "left"
const CENTER = "center"
const RIGHT = "right"
const TOP = "top"
const BOTTOM = "bottom"

export const Popup = forwardRef<PopupProps, "div">(
  (
    {
      as,
      style,
      targetElementRef,
      targetRelativePosition = {
        horizontal: "center",
        vertical: "bottom",
      },
      children,
      visible = false,
      disablePortal = false,
      flippingEnabled = false,
      onFlip,
      onOpen,
      onClose,
      ...rest
    },
    ref,
  ) => {
    const [targetElement, setTargetElement] = useState<any>()
    const [elementPosition, setElementPosition] = useState<ElementPosition>({
      x: targetElementRef?.current?.getBoundingClientRect().left,
      y: targetElementRef?.current?.getBoundingClientRect().top,
    })
    const innerRef = useRef<HTMLElement>()
    const [isOpen, setOpen] = useState(visible)
    const [
      currentTargetRelativePosition,
      setCurrentTargetRelativePosition,
    ] = useState<RelativePosition>({ ...targetRelativePosition })

    const evaluatePosition = (node = innerRef.current) => {
      const result: Position = {}

      if (targetElement) {
        if (
          currentTargetRelativePosition.horizontal === LEFT ||
          currentTargetRelativePosition.horizontal == undefined
        ) {
          result.left = String(elementPosition.x) + "px"
        } else if (currentTargetRelativePosition.horizontal === CENTER) {
          result.left =
            String(elementPosition.x + targetElement.offsetWidth / 2) + "px"
        } else if (currentTargetRelativePosition.horizontal === RIGHT) {
          result.left =
            String(elementPosition.x + targetElement.offsetWidth) + "px"
        }
        if (currentTargetRelativePosition.vertical === TOP) {
          result.top =
            String(
              elementPosition.y -
                (node?.offsetHeight !== undefined ? node.offsetHeight : 0),
            ) + "px"
        } else if (currentTargetRelativePosition.vertical === CENTER) {
          result.top =
            String(elementPosition.y + targetElement.offsetHeight / 2) + "px"
        } else if (
          currentTargetRelativePosition.vertical === BOTTOM ||
          currentTargetRelativePosition.vertical == undefined
        ) {
          result.top =
            String(elementPosition.y + targetElement.offsetHeight) + "px"
        }
      }
      return result
    }

    const [position, setPosition] = useState(evaluatePosition())

    useLayoutEffect(() => {
      setPosition(evaluatePosition())
      // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [targetElement, elementPosition, currentTargetRelativePosition])

    const callbackRef = React.useCallback((node: HTMLElement) => {
      innerRef.current = node
      if (ref) {
        if (typeof ref === "function") {
          ref(node)
        } else {
          ref.current = node
        }
      }
      if (node) {
        evaluatePosition(node)
        flip()
      }
    }, [])

    useEffect(() => {
      if (targetElementRef?.current != targetElement) {
        setTargetElement(targetElementRef?.current)
        setElementPosition({
          x: targetElementRef?.current?.getBoundingClientRect().left,
          y: targetElementRef?.current?.getBoundingClientRect().top,
        })
      }
      // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [targetElementRef?.current])

    useLayoutEffect(() => {
      if (flippingEnabled && onFlip) {
        onFlip(currentTargetRelativePosition)
      }
      // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [flippingEnabled, currentTargetRelativePosition])

    useScrollPosition(
      (prev, next) => {
        setElementPosition(next)
        flip()
      },
      targetElementRef,
      false,
      10,
    )

    useResizePosition(
      (prev, next) => {
        setElementPosition(next)
        flip()
      },
      targetElementRef,
      50,
    )

    useDragPosition(
      (prev, next) => {
        setElementPosition(next)
      },
      targetElementRef,
      false,
      10,
    )

    useEffect(() => {
      if (visible) {
        if (onOpen) onOpen()
      }
      if (visible != isOpen) {
        if (visible) {
          open()
        } else {
          close()
        }
      }
      // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [visible])

    const open = () => {
      if (!isOpen) {
        setOpen(true)
      }
      if (onOpen) {
        setTimeout(onOpen)
      }
    }

    const close = () => {
      if (isOpen) {
        setOpen(false)
        setCurrentTargetRelativePosition(targetRelativePosition)
      }
      if (onClose) {
        setTimeout(onClose)
      }
    }

    useOnClickOutside(
      targetElementRef ? [targetElementRef, innerRef] : [innerRef],
      close,
    )

    const flip = () => {
      if (!flippingEnabled || !innerRef.current) {
        return
      }
      const clientHeight = document.body.clientHeight
      const clientWidth = document.body.clientWidth

      const newPosition: RelativePosition = {}
      const popupRect = innerRef.current?.getBoundingClientRect()

      if (popupRect) {
        if (popupRect.bottom > clientHeight) {
          newPosition.vertical = "top"
        } else if (popupRect.top < 0) {
          newPosition.vertical = "bottom"
        }
        if (popupRect.right > clientWidth) {
          newPosition.horizontal = "left"
        } else if (popupRect.left < 0) {
          newPosition.horizontal = "right"
        }
      }

      setCurrentTargetRelativePosition((position) => ({
        ...position,
        ...newPosition,
      }))
    }

    if (
      isOpen &&
      (targetElementRef == undefined || targetElementRef?.current != null)
    ) {
      if (disablePortal) {
        if (as != undefined) {
          React.isValidElement(as)
          const Component = as

          return (
            <Component
              {...rest}
              style={{ ...position, ...style }}
              ref={callbackRef}
            >
              {children}
            </Component>
          )
        } else {
          return (
            <StyledDiv
              {...rest}
              style={{ ...position, ...style }}
              ref={callbackRef as any}
            >
              {children}
            </StyledDiv>
          )
        }
      } else if (as != undefined) {
        React.isValidElement(as)
        const Component = as

        return ReactDOM.createPortal(
          <Component
            {...rest}
            style={{ ...position, ...style }}
            ref={callbackRef}
          >
            {children}
          </Component>,
          document.body,
        )
      } else {
        return ReactDOM.createPortal(
          <StyledDiv
            {...rest}
            style={{ ...position, ...style }}
            ref={callbackRef as any}
          >
            {children}
          </StyledDiv>,
          document.body,
        )
      }
    } else {
      return null
    }
  },
)
