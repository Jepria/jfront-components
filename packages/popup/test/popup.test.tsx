import React from "react"
import { render } from "react-dom"
import { Popup } from "../src"
import { Link } from "@jfront/ui-link"
import { fireEvent, screen } from "@testing-library/react"

const container = document.createElement("div")
document.body.appendChild(container)

const TestComponent = () => {
  const [visible, setVisible] = React.useState(false)
  const buttonRef = React.useRef<HTMLButtonElement>(null)
  return (
    <>
      <button
        data-testid="button"
        ref={buttonRef}
        onClick={() => setVisible(!visible)}
      >
        Show
      </button>
      <Popup
        visible={visible}
        targetElementRef={buttonRef}
        targetRelativePosition={{ horizontal: "left", vertical: "bottom" }}
        onClose={() => {
          setVisible(false)
        }}
      >
        <p data-testid="content">This is popup</p>
      </Popup>
    </>
  )
}

test("Matches snapshot", () => {
  render(<TestComponent />, container)
  expect(document.body).toMatchSnapshot()
})

test("popup click", () => {
  render(<TestComponent />, container)
  fireEvent.click(screen.getByTestId("button"))
  expect(screen.getByTestId("content").innerHTML).toEqual("This is popup")
})
