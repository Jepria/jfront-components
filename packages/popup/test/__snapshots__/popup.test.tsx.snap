// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches snapshot 1`] = `
<body>
  <div>
    <button
      data-testid="button"
    >
      Show
    </button>
  </div>
</body>
`;
